thread 'coordinator' panicked at src/adapter/src/coord.rs:2324:13: error bootstrapping dataflow `as_of`

### What version of Materialize are you using?

1243449ea4118f9deb512c8f0a9f7117a6bd40ef

### What is the issue?

Checks upgrade across four versions failed in [nightly build #7697](https://buildkite.com/materialize/nightly/builds/7697#018f5565-d7e6-4d28-84e4-e83085ba29ec).

```
platform-checks-materialized-1     | thread 'coordinator' panicked at src/adapter/src/coord.rs:2324:13: error bootstrapping dataflow `as_of`: `min_as_of` [1715125812000] greater than `max_as_of` [1715125755001] (import_ids=[u692], export_ids=[u711], storage_constraints=StorageConstraints { dependencies: {User(692)}, dependants: {User(711)} })
```

--- a/src/sql/src/plan/statement/show.rs
+++ b/src/sql/src/plan/statement/show.rs
@@ -949,9 +949,6 @@ fn humanize_sql_for_show_create(
                             // Dropping a subsource does not remove any `TEXT
                             // COLUMNS` values that refer to the table it
                             // ingests, which we'll handle below.
-                            //
-                            // TODO(#26774): is this better if this is an option
-                            // on the subsources?
                             PgConfigOptionName::TextColumns => {}
                             // Drop details, which does not rountrip.
                             PgConfigOptionName::Details => return false,
@@ -984,9 +981,6 @@ fn humanize_sql_for_show_create(
                             // Dropping a subsource does not remove any `TEXT
                             // COLUMNS` values that refer to the table it
                             // ingests, which we'll handle below.
-                            //
-                            // TODO(#26774): is this better if this is an option
-                            // on the subsources?
                             MySqlConfigOptionName::TextColumns
                             | MySqlConfigOptionName::IgnoreColumns => {}
                             // Drop details, which does not rountrip.

--- a/src/storage-client/src/controller.rs
+++ b/src/storage-client/src/controller.rs
@@ -118,29 +118,6 @@ pub enum DataSource {
     Other(DataSourceOther),
 }
 
-impl DataSource {
-    /// If this data source depends on the frontier of another collection,
-    /// return it.
-    ///
-    /// Dependencies offer the ability to tie the frontiers of collections
-    /// together, e.g. we expect that you cannot drop a collection with any
-    /// dependents that have not been dropped yet.
-    pub fn collection_dependency(&self) -> Option<GlobalId> {
-        match self {
-            DataSource::Introspection(_)
-            | DataSource::Webhook
-            | DataSource::Other(DataSourceOther::TableWrites)
-            | DataSource::Progress
-            | DataSource::Other(DataSourceOther::Compute) => None,
-            // Exports depend on their ingestion. n.b. we expect this dependency
-            // to carry transitively to the ingestion's remap collection.
-            DataSource::IngestionExport { ingestion_id, .. } => Some(*ingestion_id),
-            // Ingestions depend on their remap collection.
-            DataSource::Ingestion(ingestion) => Some(ingestion.remap_collection_id),
-        }
-    }
-}
-
 /// Describes how data is written to a collection maintained outside of the
 /// storage controller.
 #[derive(Clone, Debug, Eq, PartialEq)]

--- a/src/storage-controller/src/lib.rs
+++ b/src/storage-controller/src/lib.rs
@@ -2243,7 +2243,7 @@ where
                         // do not yet track the cluster on pending compaction
                         // commands.
                         //
-                        // TODO(#24235): place the cluster ID in the pending compaction
+                        // TODO(#8185): place the cluster ID in the pending compaction
                         // commands of IngestionExports.
                         pending_source_drops.push(id);
                         None
@@ -3820,9 +3820,9 @@ impl<T: Timestamp> CollectionState<T> {
             DataSource::Webhook
             | DataSource::Introspection(_)
             | DataSource::Other(_)
-            // TODO(#24235) This isn't quite right because a source export runs on the
-            // ingestion's cluster, but we don't have the ability to perform
-            // cross-referenced lookups here (at least not yet).
+            // TODO(#8185) This isn't quite right because a source export runs
+            // on the ingestion's cluster, but we don't yet support announcing
+            // that.
             | DataSource::IngestionExport { .. }
             | DataSource::Progress => None,
         }

--- a/src/storage-controller/src/lib.rs
+++ b/src/storage-controller/src/lib.rs
@@ -1911,11 +1911,7 @@ where
         // Repeatedly extract the maximum id, and updates for it.
         while let Some(key) = updates.keys().rev().next().cloned() {
             let mut update = updates.remove(&key).unwrap();
-            if let Ok(collection) = self.collection(key) {
-                let cluster_id = self
-                    .determine_collection_cluster_id(&collection.description.data_source)
-                    .expect("collection is well structured");
-
+            if let Some(collection) = self.collections.get_mut(&key) {
                 let current_read_capabilities = collection.read_capabilities.frontier().to_owned();
                 for (time, diff) in update.iter() {
                     assert!(
@@ -1938,11 +1934,6 @@ where
                     }
                 }
 
-                let collection = self
-                    .collections
-                    .get_mut(&key)
-                    .expect("collection must still exists");
-
                 let changes = collection.read_capabilities.update_iter(update.drain());
                 update.extend(changes);
 
@@ -1953,9 +1944,14 @@ where
                         .extend(update.iter().cloned());
                 }
 
-                let (changes, frontier, _cluster_id) = collections_net
-                    .entry(key)
-                    .or_insert_with(|| (ChangeBatch::new(), Antichain::new(), cluster_id));
+                let (changes, frontier, _cluster_id) =
+                    collections_net.entry(key).or_insert_with(|| {
+                        (
+                            ChangeBatch::new(),
+                            Antichain::new(),
+                            collection.cluster_id(),
+                        )
+                    });
 
                 changes.extend(update.drain());
                 *frontier = collection.read_capabilities.frontier().to_owned();
@@ -2241,6 +2237,17 @@ where
                         };
                         Some(drop_fut.boxed())
                     }
+                    DataSource::IngestionExport { .. } if read_frontier.is_empty() => {
+                        // Dropping an ingestion is a form of dropping a source.
+                        // This won't be handled above because ingestion exports
+                        // do not yet track the cluster on pending compaction
+                        // commands.
+                        //
+                        // TODO(#24235): place the cluster ID in the pending compaction
+                        // commands of IngestionExports.
+                        pending_source_drops.push(id);
+                        None
+                    }
                     // These sources are manged by `clusterd`.
                     DataSource::Webhook
                     | DataSource::Introspection(_)
@@ -2418,12 +2425,10 @@ where
 
         // Enrich `frontiers` with storage frontiers.
         for (object_id, collection) in self.collections.iter().filter(|(_id, c)| !c.is_dropped()) {
-            let replica_id = self
-                .determine_collection_cluster_id(&collection.description.data_source)
-                .expect("collection is well structured")
+            let replica_id = collection
+                .cluster_id()
                 .and_then(|c| self.replicas.get(&c))
                 .copied();
-
             if let Some(replica_id) = replica_id {
                 let upper = collection.write_frontier.clone();
                 frontiers.insert((*object_id, replica_id), upper);
@@ -3551,34 +3556,6 @@ where
         Ok(dependency)
     }
 
-    /// Determine which, if any, cluster this `DataSource` runs on.
-    fn determine_collection_cluster_id(
-        &self,
-        data_source: &DataSource,
-    ) -> Result<Option<StorageInstanceId>, StorageError<T>> {
-        let instance_id = match data_source {
-            DataSource::Ingestion(ingestion) => Some(ingestion.instance_id),
-            DataSource::IngestionExport { ingestion_id, .. } => {
-                let ingestion_collection = self.collection(*ingestion_id)?;
-                match &ingestion_collection.description.data_source {
-                    DataSource::Ingestion(i) => Some(i.instance_id),
-                    _ => {
-                        tracing::error!(
-                            "IngestionExport relies on {ingestion_id}, which is not an ingestion"
-                        );
-                        Err(StorageError::IdentifierInvalid(*ingestion_id))?
-                    }
-                }
-            }
-            DataSource::Webhook
-            | DataSource::Introspection(_)
-            | DataSource::Other(_)
-            | DataSource::Progress => None,
-        };
-
-        Ok(instance_id)
-    }
-
     /// If this identified collection has a dependency, install a read hold on
     /// it.
     ///
@@ -3833,6 +3810,21 @@ impl<T: Timestamp> CollectionState<T> {
             storage_dependency,
             write_frontier,
             collection_metadata: metadata,
+        }
+    }
+
+    /// Returns the cluster to which the collection is bound, if applicable.
+    pub fn cluster_id(&self) -> Option<StorageInstanceId> {
+        match &self.description.data_source {
+            DataSource::Ingestion(ingestion) => Some(ingestion.instance_id),
+            DataSource::Webhook
+            | DataSource::Introspection(_)
+            | DataSource::Other(_)
+            // TODO(#24235) This isn't quite right because a source export runs on the
+            // ingestion's cluster, but we don't have the ability to perform
+            // cross-referenced lookups here (at least not yet).
+            | DataSource::IngestionExport { .. }
+            | DataSource::Progress => None,
         }
     }
 

--- a/src/storage-controller/src/lib.rs
+++ b/src/storage-controller/src/lib.rs
@@ -637,7 +637,8 @@ where
             let data_shard_since = since_handle.since().clone();
 
             // Determine if this collection has another dependency.
-            let storage_dependency = description.data_source.collection_dependency();
+            let storage_dependency =
+                self.determine_collection_dependency(&description.data_source)?;
 
             // Determine the intial since of the collection.
             let initial_since = match storage_dependency {
@@ -3514,6 +3515,40 @@ where
             self.finalizable_shards.remove(&id);
             self.finalized_shards.insert(id);
         }
+    }
+
+    /// Determine if this collection has another dependency.
+    ///
+    /// Currently, collections have either 0 or 1 dependencies.
+    fn determine_collection_dependency(
+        &self,
+        data_source: &DataSource,
+    ) -> Result<Option<GlobalId>, StorageError<T>> {
+        let dependency = match &data_source {
+            DataSource::Introspection(_)
+            | DataSource::Webhook
+            | DataSource::Other(DataSourceOther::TableWrites)
+            | DataSource::Progress
+            | DataSource::Other(DataSourceOther::Compute) => None,
+            DataSource::IngestionExport { ingestion_id, .. } => {
+                // Ingestion exports depend on their primary source's remap
+                // collection.
+                let source_collection = self.collection(*ingestion_id)?;
+                match &source_collection.description {
+                    CollectionDescription {
+                        data_source: DataSource::Ingestion(ingestion_desc),
+                        ..
+                    } => Some(ingestion_desc.remap_collection_id),
+                    _ => unreachable!(
+                        "SourceExport must only refer to primary sources that already exist"
+                    ),
+                }
+            }
+            // Ingestions depend on their remap collection.
+            DataSource::Ingestion(ingestion) => Some(ingestion.remap_collection_id),
+        };
+
+        Ok(dependency)
     }
 
     /// Determine which, if any, cluster this `DataSource` runs on.

--- a/src/storage-controller/src/rehydration.rs
+++ b/src/storage-controller/src/rehydration.rs
@@ -494,8 +494,7 @@ where
                         Some(export) => {
                             export.description.as_of.clone_from(frontier);
                         }
-                        // uppers contains both ingestions and their exports
-                        None if self.uppers.contains_key(id) => continue,
+                        None if self.sources.contains_key(id) => continue,
                         None => panic!("AllowCompaction command for non-existent {id}"),
                     }
                 }

--- a/src/storage/src/storage_state.rs
+++ b/src/storage/src/storage_state.rs
@@ -965,51 +965,16 @@ impl<'w, A: Allocate> Worker<'w, A> {
                 }
                 StorageCommand::RunIngestions(ingestions) => {
                     ingestions.retain_mut(|ingestion| {
-                        // Subsources can be dropped independently of their
-                        // primary source, so we evaluate them in a separate
-                        // loop.
-                        for export_id in ingestion
-                            .description
-                            .source_exports
-                            .keys()
-                            .filter(|export_id| **export_id != ingestion.id)
-                        {
-                            if drop_commands.remove(export_id)
-                                || self.storage_state.dropped_ids.contains(&ingestion.id)
-                            {
-                                self.storage_state.dropped_ids.insert(*export_id);
-                            }
-                        }
-
                         if drop_commands.remove(&ingestion.id)
                             || self.storage_state.dropped_ids.contains(&ingestion.id)
                         {
-                            // If an ingestion is dropped, so too must all of
-                            // its subsources (i.e. ingestion exports, as well
-                            // as its progress subsource).
-                            for id in ingestion.description.subsource_ids() {
-                                drop_commands.remove(&id);
-                                self.storage_state.dropped_ids.insert(id);
-                            }
+                            // Make sure that we report back that the ID was
+                            // dropped.
+                            self.storage_state.dropped_ids.insert(ingestion.id);
 
                             false
                         } else {
-                            let most_recent_defintion =
-                                seen_most_recent_definition.insert(ingestion.id);
-
-                            if most_recent_defintion {
-                                // If this is the most recent definition, this
-                                // is what we will be running when
-                                // reconciliation completes. This definition
-                                // must not include any dropped subsources.
-                                ingestion.description.source_exports.retain(|export_id, _| {
-                                    !self.storage_state.dropped_ids.contains(export_id)
-                                });
-
-                                // After clearing any dropped subsources, we can
-                                // state that we expect all of these to exist.
-                                expected_objects.extend(ingestion.description.subsource_ids());
-                            }
+                            expected_objects.extend(ingestion.description.subsource_ids());
 
                             let running_ingestion =
                                 self.storage_state.ingestions.get(&ingestion.id);
@@ -1019,7 +984,7 @@ impl<'w, A: Allocate> Worker<'w, A> {
                             //   is why these commands are run in reverse.
                             // - Ingestions whose descriptions are not exactly
                             //   those that are currently running.
-                            most_recent_defintion
+                            seen_most_recent_definition.insert(ingestion.id)
                                 && running_ingestion != Some(&ingestion.description)
                         }
                     })
@@ -1072,14 +1037,12 @@ impl<'w, A: Allocate> Worker<'w, A> {
         let stale_objects = self
             .storage_state
             .ingestions
-            .values()
-            .map(|i| i.subsource_ids())
-            .flatten()
-            .chain(self.storage_state.exports.keys().copied())
+            .keys()
+            .chain(self.storage_state.exports.keys())
             // Objects are considered stale if we did not see them re-created.
             .filter(|id| !expected_objects.contains(id))
             // Synthesize the drop command
-            .map(|id| (id, Antichain::new()))
+            .map(|id| (*id, Antichain::new()))
             .collect::<Vec<_>>();
 
         trace!(

--- a/src/storage/src/storage_state.rs
+++ b/src/storage/src/storage_state.rs
@@ -1192,9 +1192,7 @@ impl StorageState {
                             // restart a sink in the future.
                             export_description.as_of.clone_from(&frontier);
                         }
-                        // reported_frontiers contains both ingestions and their
-                        // exports
-                        None if self.reported_frontiers.contains_key(&id) => (),
+                        None if self.ingestions.contains_key(&id) => (),
                         None => panic!("AllowCompaction command for non-existent {id}"),
                     }
 

